From 1c771c6e40c5343eabafba0134dc67a1c5625b10 Mon Sep 17 00:00:00 2001
From: BodoBolero <peterbendel@neon.tech>
Date: Thu, 23 May 2024 09:49:35 +0200
Subject: [PATCH] From: Heikki Linnakangas <heikki.linnakangas@iki.fi> Date:
 Fri, 2 Feb 2024 22:26:45 +0200 Subject: [PATCH 1/1] Make v0.6.0 work with
 Neon

Now that the WAL-logging happens as a separate step at the end of the
build, we need a few neon-specific hints to make it work.

Peter Bendel: also remove `-march=native`from OPTFLAGS to get a compatible binary
---
 Makefile        |  2 +-
 src/hnswbuild.c | 36 ++++++++++++++++++++++++++++++++++++
 2 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7684684..5a56c17 100644
--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ REGRESS = $(patsubst test/sql/%.sql,%,$(TESTS))
 REGRESS_OPTS = --inputdir=test --load-extension=$(EXTENSION)
 
 # To compile for portability, run: make OPTFLAGS=""
-OPTFLAGS = -march=native
+OPTFLAGS=""
 
 # Mac ARM doesn't always support -march=native
 ifeq ($(shell uname -s), Darwin)
diff --git a/src/hnswbuild.c b/src/hnswbuild.c
index ee94ed9..b3d2fcf 100644
--- a/src/hnswbuild.c
+++ b/src/hnswbuild.c
@@ -860,9 +860,17 @@ HnswParallelBuildMain(dsm_segment *seg, shm_toc *toc)
 
 	hnswarea = shm_toc_lookup(toc, PARALLEL_KEY_HNSW_AREA, false);
 
+#ifdef NEON_SMGR
+	smgr_start_unlogged_build(RelationGetSmgr(indexRel));
+#endif
+
 	/* Perform inserts */
 	HnswParallelScanAndInsert(heapRel, indexRel, hnswshared, hnswarea, false);
 
+#ifdef NEON_SMGR
+	smgr_finish_unlogged_build_phase_1(RelationGetSmgr(indexRel));
+#endif
+
 	/* Close relations within worker */
 	index_close(indexRel, indexLockmode);
 	table_close(heapRel, heapLockmode);
@@ -1117,13 +1125,41 @@ BuildIndex(Relation heap, Relation index, IndexInfo *indexInfo,
 	SeedRandom(42);
 #endif
 
+#ifdef NEON_SMGR
+	smgr_start_unlogged_build(RelationGetSmgr(index));
+#endif
+
 	InitBuildState(buildstate, heap, index, indexInfo, forkNum);
 
 	BuildGraph(buildstate, forkNum);
 
+#ifdef NEON_SMGR
+	smgr_finish_unlogged_build_phase_1(RelationGetSmgr(index));
+#endif
+
 	if (RelationNeedsWAL(index))
+	{
 		log_newpage_range(index, forkNum, 0, RelationGetNumberOfBlocks(index), true);
 
+#ifdef NEON_SMGR
+		{
+#if PG_VERSION_NUM >= 160000
+			RelFileLocator rlocator = RelationGetSmgr(index)->smgr_rlocator.locator;
+#else
+			RelFileNode rlocator = RelationGetSmgr(index)->smgr_rnode.node;
+#endif
+
+			SetLastWrittenLSNForBlockRange(XactLastRecEnd, rlocator,
+										   MAIN_FORKNUM, 0, RelationGetNumberOfBlocks(index));
+			SetLastWrittenLSNForRelation(XactLastRecEnd, rlocator, MAIN_FORKNUM);
+		}
+#endif
+	}
+
+#ifdef NEON_SMGR
+	smgr_end_unlogged_build(RelationGetSmgr(index));
+#endif
+
 	FreeBuildState(buildstate);
 }
 
-- 
2.39.3 (Apple Git-146)

